### ISSUE TITLE
When crosscompiling to Windows from Linux build crashes

### ISSUE THREAD

**Issue by aido93** [randombit/botan#1548]
Hello!

I try to build libbotan using this flags:

`./configure.py --os=mingw --with-openssl --with-zlib --cc-bin=/home/builder/mxe/usr/bin/i686-w64-mingw32.static-g++ --with-external-includedir=/home/builder/mxe/usr/i686-w64-mingw32.static/include --without-os-features=threads --with-external-libdir=/home/builder/mxe/usr/i686-w64-mingw32.static/lib --cpu=i686`

And I have these linkage errors:
`/home/builder/mxe/usr/bin/i686-w64-mingw32.static-g++ -fstack-protector -pthread -L/home/builder/mxe/usr/i686-w64-mingw32.static/lib  build/obj/cli/asn1.o build/obj/cli/cc_enc.o build/obj/cli/cli.o build/obj/cli/cli_rng.o build/obj/cli/compress.o build/obj/cli/encryption.o build/obj/cli/main.o build/obj/cli/math.o build/obj/cli/psk.o build/obj/cli/pubkey.o build/obj/cli/speed.o build/obj/cli/timing_tests.o build/obj/cli/tls_client.o build/obj/cli/tls_http_server.o build/obj/cli/tls_proxy.o build/obj/cli/tls_server.o build/obj/cli/tls_utils.o build/obj/cli/utils.o build/obj/cli/x509.o -L. -lbotan-2 -lssl -lcrypto -lws2_32 -lz -lgdi32 -o botan.exe
build/obj/cli/encryption.o:encryption.cpp:(.text$_ZN9Botan_CLI10Encryption2goEv[__ZN9Botan_CLI10Encryption2goEv]+0x763): undefined reference to `Botan::Cipher_Mode::create(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, Botan::Cipher_Dir, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
build/obj/cli/speed.o:speed.cpp:(.text$_ZN9Botan_CLI5Speed12bench_mp_mulENSt6chrono8durationIxSt5ratioILx1ELx1000EEEE[__ZN9Botan_CLI5Speed12bench_mp_mulENSt6chrono8durationIxSt5ratioILx1ELx1000EEEE]+0x54d): undefined reference to `Botan::BigInt::square(std::vector<unsigned int, Botan::secure_allocator<unsigned int> >&)'
build/obj/cli/speed.o:speed.cpp:(.text$_ZN9Botan_CLI5Speed2goEv[__ZN9Botan_CLI5Speed2goEv]+0x19da): undefined reference to `Botan::Cipher_Mode::create(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, Botan::Cipher_Dir, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
build/obj/cli/speed.o:speed.cpp:(.text$_ZN9Botan_CLI5Speed2goEv[__ZN9Botan_CLI5Speed2goEv]+0x1a53): undefined reference to `Botan::Cipher_Mode::create_or_throw(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, Botan::Cipher_Dir, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
build/obj/cli/timing_tests.o:timing_tests.cpp:(.text+0xf6d): undefined reference to `Botan::Cipher_Mode::create(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, Botan::Cipher_Dir, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
build/obj/cli/timing_tests.o:timing_tests.cpp:(.text+0x38eb): undefined reference to `Botan::TLS::TLS_CBC_HMAC_AEAD_Mode::TLS_CBC_HMAC_AEAD_Mode(Botan::Cipher_Dir, std::unique_ptr<Botan::BlockCipher, std::default_delete<Botan::BlockCipher> >, std::unique_ptr<Botan::MessageAuthenticationCode, std::default_delete<Botan::MessageAuthenticationCode> >, unsigned int, unsigned int, bool, bool)'
build/obj/cli/timing_tests.o:timing_tests.cpp:(.text+0x3c58): undefined reference to `Botan::TLS::TLS_CBC_HMAC_AEAD_Mode::TLS_CBC_HMAC_AEAD_Mode(Botan::Cipher_Dir, std::unique_ptr<Botan::BlockCipher, std::default_delete<Botan::BlockCipher> >, std::unique_ptr<Botan::MessageAuthenticationCode, std::default_delete<Botan::MessageAuthenticationCode> >, unsigned int, unsigned int, bool, bool)'
build/obj/cli/timing_tests.o:timing_tests.cpp:(.text+0x4164): undefined reference to `Botan::TLS::TLS_CBC_HMAC_AEAD_Mode::TLS_CBC_HMAC_AEAD_Mode(Botan::Cipher_Dir, std::unique_ptr<Botan::BlockCipher, std::default_delete<Botan::BlockCipher> >, std::unique_ptr<Botan::MessageAuthenticationCode, std::default_delete<Botan::MessageAuthenticationCode> >, unsigned int, unsigned int, bool, bool)'
collect2: error: ld returned 1 exit status`

It seems that there are some errors in test applications (botan.exe)

**Comment by randombit** [comment]
You may want to try a clean rebuild (which can be done by re-running configure). It looks like you have a partial rebuild: all of the functions listed (`BigInt::square`, `TLS_CBC_HMAC_AEAD_Mode` etc) are either recently added or recently changed signatures.

**Comment by aido93** [comment]
Nope. Right now I tried to clone and build master branch. And errors are the same

**Comment by randombit** [comment]
I cannot replicate with 

```
Using built-in specs.
COLLECT_GCC=i686-w64-mingw32-g++
COLLECT_LTO_WRAPPER=/usr/lib/gcc/i686-w64-mingw32/7.2.0/lto-wrapper
Target: i686-w64-mingw32
Configured with: /build/mingw-w64-gcc/src/gcc/configure --prefix=/usr --libexecdir=/usr/lib --target=i686-w64-mingw32 --enable-languages=c,lto,c++,objc,obj-c++,fortran,ada --enable-shared --enable-static --enable-threads=posix --enable-fully-dynamic-string --enable-libstdcxx-time=yes --with-system-zlib --enable-cloog-backend=isl --enable-lto --disable-dw2-exceptions --enable-libgomp --disable-multilib --enable-checking=release
Thread model: posix
gcc version 7.2.0 (GCC) 
```

**Comment by randombit** [comment]
Do you see the symbols if you dump with `nm`?

```
$ i686-w64-mingw32-nm -C build/obj/lib/math_bigint_big_ops2.o | grep square
00001240 T Botan::BigInt::square(std::vector<unsigned int, Botan::secure_allocator<unsigned int> >&)
```

**Comment by aido93** [comment]
Oh... I changed compiler to 6.3.0 and there are a lot of warnings but ... errors still present :(

**Comment by aido93** [comment]
$ make tests

`build/obj/test/test_aead.o:test_aead.cpp:(.text+0x74b): undefined reference to `Botan::AEAD_Mode::create(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, Botan::Cipher_Dir, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
build/obj/test/test_aead.o:test_aead.cpp:(.text+0x24b6): undefined reference to `Botan::AEAD_Mode::create(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, Botan::Cipher_Dir, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
build/obj/test/test_aead.o:test_aead.cpp:(.text+0x52b4): undefined reference to `Botan::AEAD_Mode::create(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, Botan::Cipher_Dir, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
build/obj/test/test_aead.o:test_aead.cpp:(.text+0x5308): undefined reference to `Botan::AEAD_Mode::create(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, Botan::Cipher_Dir, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
build/obj/test/test_dlies.o:test_dlies.cpp:(.text+0x1614): undefined reference to `Botan::Cipher_Mode::create(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, Botan::Cipher_Dir, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
build/obj/test/test_dlies.o:test_dlies.cpp:(.text+0x1674): undefined reference to `Botan::Cipher_Mode::create(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, Botan::Cipher_Dir, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
build/obj/test/test_filters.o:test_filters.cpp:(.text$_ZN11Botan_Tests12Filter_Tests13test_pipe_cbcEv[__ZN11Botan_Tests12Filter_Tests13test_pipe_cbcEv]+0x17d): undefined reference to `Botan::Cipher_Mode::create(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, Botan::Cipher_Dir, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
build/obj/test/test_filters.o:test_filters.cpp:(.text$_ZN11Botan_Tests12Filter_Tests13test_pipe_cbcEv[__ZN11Botan_Tests12Filter_Tests13test_pipe_cbcEv]+0xb43): undefined reference to `Botan::Cipher_Mode::create(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, Botan::Cipher_Dir, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
build/obj/test/test_modes.o:test_modes.cpp:(.text$_ZN11Botan_Tests26Cipher_Mode_IV_Carry_Tests17test_cbc_iv_carryEv[__ZN11Botan_Tests26Cipher_Mode_IV_Carry_Tests17test_cbc_iv_carryEv]+0x1a5): undefined reference to `Botan::Cipher_Mode::create(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, Botan::Cipher_Dir, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
build/obj/test/test_modes.o:test_modes.cpp:(.text$_ZN11Botan_Tests26Cipher_Mode_IV_Carry_Tests17test_cbc_iv_carryEv[__ZN11Botan_Tests26Cipher_Mode_IV_Carry_Tests17test_cbc_iv_carryEv]+0x297): more undefined references to `Botan::Cipher_Mode::create(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, Botan::Cipher_Dir, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)' follow
build/obj/test/test_tls.o:test_tls.cpp:(.text$_ZN11Botan_Tests13TLS_CBC_Tests12run_one_testERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEERKSt13unordered_mapIS6_S6_St4hashIS6_ESt8equal_toIS6_ESaISt4pairIS7_S6_EEE[__ZN11Botan_Tests13TLS_CBC_Tests12run_one_testERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEERKSt13unordered_mapIS6_S6_St4hashIS6_ESt8equal_toIS6_ESaISt4pairIS7_S6_EEE]+0x3f9): undefined reference to `Botan::TLS::TLS_CBC_HMAC_AEAD_Mode::TLS_CBC_HMAC_AEAD_Mode(Botan::Cipher_Dir, std::unique_ptr<Botan::BlockCipher, std::default_delete<Botan::BlockCipher> >, std::unique_ptr<Botan::MessageAuthenticationCode, std::default_delete<Botan::MessageAuthenticationCode> >, unsigned int, unsigned int, bool, bool)'
build/obj/test/unit_ecc.o:unit_ecc.cpp:(.text+0x59ef): undefined reference to `Botan::EC_Group::a_is_zero() const'
collect2: error: ld returned 1 exit status`

**Comment by randombit** [comment]
Maybe try 2.6.0 release?

**Comment by aido93** [comment]
Same errors were

**Comment by aido93** [comment]
Also.. Could you fix some versions of compilers because server software often doesn't change for years?

**Comment by randombit** [comment]
> Also.. Could you fix some versions of compilers because server software often doesn't change for years?

I do not understand the request. We have an explicit list of which platforms and compiler versions we support https://botan.randombit.net/manual/support.html

**Comment by aido93** [comment]
Now I don't understand too =) Cause my mingw-compiler satisfies your requirements

**Comment by randombit** [comment]
Indeed. And MinGW GCC works fine for me locally, and also in CI. So I really don't have any idea why it is not working for you.

**Comment by aido93** [comment]
Ok. maybe you have already packed libraries and includes? Something like rpm or deb package or just tar.gz?

**Comment by aido93** [comment]
Also. How libbotan links with libssl? I don't see it in Makefile but I set --with-openssl flag.

**Comment by aido93** [comment]
Resolved.

Please change line with EXE_LINKS_TO variable in Makefile to this:
`EXE_LINKS_TO   = ./libbotan-2.a $(LIB_LINKS_TO)`

[see here for details](https://stackoverflow.com/questions/26392525/static-library-linked-but-still-undefined-references)

**Comment by randombit** [comment]
I don't see how the SO link is relevant - we are not doing `-L/path/to/lib.a` (which certainly is incorrect). `-L. -lbotan-2` and `./libbotan-2.a` should have the same effect.

**Comment by aido93** [comment]
Also when I cross-compile library it does not appear a dll-version of library. But I didn't see into the problem.
